reco_utils/dataset/pandas_df_utils.py

```diff
@@ -117,55 +117,120 @@ def libffm_converter(df, col_rating=DEFAULT_RATING_COL, filepath=None):
-                # If not, put them into the key-values of the dict and count the index.
-                if (field, feature) not in field_feature_dict:
-                    field_feature_dict[(field, feature)] = idx
-                    idx += 1
-
-    def _convert(field, feature, field_index, field_feature_index_dict):
-        if isinstance(feature, str):
-            field_feature_index = field_feature_index_dict[(field, feature)]
-            feature = 1
-        else:
-            field_feature_index = field_index
-
-        return "{}:{}:{}".format(field_index, field_feature_index, feature)
-
-    for col_index, col in enumerate(field_names):
-        df_new[col] = df_new[col].apply(lambda x: _convert(col, x, col_index+1, field_feature_dict))
-
-    # Move rating column to the first.
-    field_names.insert(0, col_rating)
-    df_new = df_new[field_names]
-
-    if filepath is not None:
-        np.savetxt(filepath, df_new.values, delimiter=' ', fmt='%s')
-
-    return df_new
-
+    def __init__(self, filepath=None):
+        if filepath is not None:
+            # Check the existence of path
+            import os
+            if not os.path.exists(filepath):
+                raise ValueError("The specified file path {} does not exist".format(self.filepath))
+        
+        self.filepath = filepath
+    
+    def fit(self, df, col_rating=DEFAULT_RATING_COL):
+        """Fit the dataframe for libffm format. In there method does nothing but check the validity of 
+        the input columns
+
+        Args:
+            df (pd.DataFrame): input Pandas dataframe.
+            col_rating (str): rating of the data.
+
+        Return:
+            The instance of the conveter
+        """
+        # Check column types.
+        types = df.dtypes
+        if not all([x == object or np.issubdtype(x, np.integer) or x == np.float for x in types]):
+            raise TypeError("Input columns should be only object and/or numeric types.")
+
+        if col_rating not in df.columns:
+            raise TypeError("Column of {} is not in input dataframe columns".format(col_rating))
+
+        self.col_rating = col_rating
+        self.field_names = list(df.drop(col_rating, axis=1).columns)
+
+        return self
+
+    def transform(self, df):
+        """Tranform an input dataset with the same schema (column names and dtypes) to libffm format 
+        by using the fitted converter.
+
+        Args: 
+            df (pd.DataFrame): input Pandas dataframe.
+
+        Return:
+            pd.DataFrame: output libffm format dataframe.
+        """
+        if not self.col_rating in df.columns:
+            raise ValueError("Input dataset does not contain the label column {} in the fitting dataset".format(self.col_rating))
+
+        if not all([x in df.columns for x in self.field_names]):
+            raise ValueError("Not all columns in the input dataset appear in the fitting dataset")
+
+        df_new = df.copy()
```

Comment:
is there a way to avoid this to reduce memory? In my mind we are taking data from the input df and convert it in another thing. Are there parts of df that we are reusing or the output is completely new?

Reply:
Removed it.